core/arch/arm/mm/core_mmu.c

```diff
@@ -842,49 +900,118 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
+	 * all small-page mapped area before the pgdir mapped areas.
+	 */
+	qsort(memory_map, last, sizeof(struct tee_mmap_region),
+		cmp_mmap_by_smaller_region_size);
+
+	/* Flat map and vaspace have defined a reserved virtual range */
+	vaspace_start = va;
+	vaspace_size = end - va;
+
+	/*
+	 * If coherent and vaspace do not use the same pgdir entry/ies, it
+	 * is likely that we do not execute in a small super secure memory
+	 * but rather in a bigger RAM without pager. Therefore do not
+	 * optimize xlat reuse and reuse only around vaspace.
+	 * If coherent and vaspace use the same pgdir entry/ies, optimize
+	 * xlat reuse around coherent and vaspace. Since optee_os pager
+	 * mapping, it is likely that coherent memory is below unpaged
+	 * sections.
+	 */
+	pgdir_sharing =
+		vaspace_coherent_share_pgdir(vaspace_start, vaspace_size,
+					     coherent_start, coherent_size);
+
+	if (pgdir_sharing && coherent_start < vaspace_start) {
+		/* Resuse xlat entries before coherent */
+		vstart = ROUNDDOWN(coherent_start, CORE_MMU_PGDIR_SIZE);
+		init_smallpage_map(memory_map, vstart, coherent_start, false);
+
+		vstart = coherent_start + coherent_size;
+	} else
+		vstart = ROUNDDOWN(vaspace_start, CORE_MMU_PGDIR_SIZE);
+
+	/* Resuse xlat entries before vaspace */
+	init_smallpage_map(memory_map, vstart, vaspace_start, false);
+	vstart = vaspace_start + vaspace_size;
+
+	if (pgdir_sharing && coherent_start > vaspace_start) {
+		/* Resuse xlat entries between vaspace and coherent */
+		init_smallpage_map(memory_map, vstart, coherent_start, false);
+
+		vstart = coherent_start + coherent_size;
+	}
+
+	/* Resuse xlat entries above coherent/vaspace */
+	vstart = init_smallpage_map(memory_map, vstart, ~0, false);
 
 #if !defined(CFG_WITH_LPAE)
-			if (va_is_secure != map_is_secure(map)) {
-				va_is_secure = !va_is_secure;
-				va = ROUNDDOWN(va, CORE_MMU_PGDIR_SIZE);
-			}
+	/* 2nd level xlat for non-secure small page mapping */
+	vstart = ROUNDUP(vstart, CORE_MMU_PGDIR_SIZE);
+	vstart = init_smallpage_map(memory_map, vstart, ~0, true);
 #endif
-			map->attr = core_mmu_type_to_attr(map->type);
-			va -= map->size;
-			va = ROUNDDOWN(va, map->region_size);
-#if !defined(CFG_WITH_LPAE)
-			/* Mapping does not yet support sharing L2 tables */
-			va = ROUNDDOWN(va, CORE_MMU_PGDIR_SIZE);
+
+	/* Vaspace now covers the assigned virtual ranges */
+	if (!pgdir_sharing)
+		vaspace_start = ROUNDDOWN(vaspace_start, CORE_MMU_PGDIR_SIZE);
+	else
+		vaspace_start = ROUNDDOWN(MIN(vaspace_start, coherent_start),
+					  CORE_MMU_PGDIR_SIZE);
+	vaspace_size = ROUNDUP(vstart, CORE_MMU_PGDIR_SIZE) - vaspace_start;
+
+	/* All small page map area shall be assigned a location */
+	for (map = memory_map; !core_mmap_is_end_of_table(map); map++)
+		assert(map->va || map->region_size != SMALL_PAGE_SIZE);
+
+#ifdef CFG_WITH_LPAE
+	/* LPAE: 1 pgdir for protection, user memory located above the core */
+	va = CORE_MMU_PGDIR_SIZE;
```

Comment:
> So we start to map the rest at 2 MiB regardless of where flat mapped range is (it could even conflict with this)? 

The loop right after this init skips the virtual range already used to assigned virtual locations. This virtual area is expected to be fully cover the current value of `vaspace_start`/`vaspace_size`.

> More efficient use of level 2 tables would be to keep in same level 2 table as used above, it's 1 GiB so there should be plenty of space.

Agree. I will use the full 1GB around the flat map area to reuse the level2 xlat table.